### PR TITLE
Waterworks: Plumbing enabled sinks, tanks, plant trays, and showers.

### PR DESCRIPTION
**Pull request by ArcaneMusic**
## About The Pull Request

God, I can't tell if this is more horrifying than incredible. Lets settle on incredible until I start getting DM's about the horrible evil I've unleashed.
**https://cdn.discordapp.com/attachments/326831214667235328/740812723956482118/2020-08-06_02-04-14.mp4
_For everything in this PR working at the same time, watch this video!_**

Alright, from the top:
### Sinks now need plumbing to pump out liquids.
All sinks now has a demand-only plumbing pipe face the opposite direction of the sink itself. A standard iron sink will hold 100 units of water, and must be constructed from a sink frame. To complete the frame, a new stock part, the water recycler must be applied to the frame. Once the sink is constructed, it will only hold those 100 units of water until new water is piped in.
This however doesn't extend just to water, as a result. Any reagent can be piped into the sink, and will be used 5 units at a time per use. This means water isn't just an infinite resource, and will need to be resupplied to the station if, somehow, you're using a metric ton of water somehow.
image
_A basic sink setup with water being plumbed in._
### Showers now need plumbing to wash you off.
Want to dispense reagents faster on your ~~Patients~~ ~~Victims~~ ~~Friends~~ Coworkers? Well, you may also construct a shower frame, in a similar fashion to a sink frame, and pipe in your own reagents into the loop. The shower drains reagents much faster than the shower, and also preforms vapor and touch reagent reactions, resulting in a constant stream of chemicals being processed. This can lead to horrible situations, like a chemical heater hooked up to a vat of cooking oil being pumped into a shower.
Additionally, showers will display the reagents they're pumping out visually based on the color of their chemical contents. As a result, dangerous or unique showers are going to look noticeably different if their contents are dangerous looking.
image
_A safe shower._
image
_A very unsafe shower._
### Wait what was that bit about cooking oil?
So, reagent dispensers, like water/fuel tanks, Virus Food, Cooking Oil Vats, Pepperspray, and the like, can now be converted into a stationary tank of reagents by using a sheet of metal on it, allowing it to behave as a simple supply of their respective contents, and can be pumped into a plumbing network. This means rolling water tanks can be used to resupply sinks and showers, as well as allow for inventive uses of otherwise unused chemicals located all around the station. Pump welding fuel into a resupply line to fight a blob! Pump pepper spray into the dorms showers! Create a deep fryer shower! The possibilities are LIMITED (Because we have very few reagent dispensers)! Still, this will prevent sinks and showers from being completely unfix-able should all the water run out, as cargo can purchase many of these chemical tanks at will.

### Hydroponics is now also plumbing enabled.
So there I was, thinking, have I gone mad? Is the barking I hear in my noggin my own dogs rattling and raging against the eternal ocean tide, or are these the hounds of tindalos come to finally rip me to shreds within the sea of causality?
The answer to this is quite simple: I added plumbing simple-demand components to hydroponics.
Now, the trays will accept a pumping input, which can provide the plants themselves a reagent input for their nutrients, and SHOULD (current not fixed yet) also contribute to refilling the trays water. As a check to this horrible decision, there are very few plumbing chems that don't suffer from some kind of downside if exclusively pumped into the trays, so this should be done delicately and with forethought applied, or you'll have a row of plants with dead stats because they OD'd on ammonia.
image

## Holy damn why are you doing this
I'm thinking this pr is probably in need of some serious trimming and will probably be closed as a result of the need to atomize this down, but this way I can start to get some feedback and ideas on which aspects would be better suited towards the current design direction of plumbing/bio-chem/reagent heavy jobs. As far as I'm aware, this is all in accordance with the relevant design docs, but this way I'll know for sure 🤷‍♂️ 

## To-do before being 100% sparkly and clean

- [x] Fix hydroponics trays adding water to the nutrient reagents slot (NOT A TRUE FIX, SEE COMMENTS BELOW PLEASE)
- [x] Make disconnected sinks/showers slowly regenerate water so that a stationwide plumbing net isn't required so Anturk doesn't snap my kneecaps off
- [x] Make water/fuel tanks use their snowflaked stationary tank sprites instead of the custom overlay version
- [x] Yeet out the existing psudo-plumbing from hydroponics

## Why It's Good For The Game
We've discussed adding plumbing into the game for forever now, and with all the new and tangentally related plumbing content we have, almost everything we have now perfectly enables servers to properly create a station-wide plumbing network.

Sinks and showers provide an infinite supply of water all shift long, and since they can be produced with any material, anywhere, practically for free, that was the first and foremost reason why I made this PR, to make water slightly more precious on station, to justify the 20+ water tanks mapped across every station. Plus, no more meatsink spam in the middle of the hallways.

Hydroponics has a built in feature that was intended to work similar to plumbing, before we had plumbing, for sharing resources/water, but it was pretty rough and is still exceptionally difficult to make look very pretty on live servers. By integrating them with actual plumbing pipes, this enables one step closer to integrating chemistry and botany to merge going forward, and give them an incentive to include each other in their supply chains each shift.

On the reagent tank conversion, That's just an interesting feature I felt the need for while testing everything else, and found pretty great success in. Being able to pipe in large supplies of a single reagent that ordinarily would just sit on a wall or in mait somewhere proved extra useful, and would be beneficial for refilling plumbing heavy areas like bathrooms/medbay, etc.

## Changelog
:cl:
add: Sinks and showers can now have reagents piped into them.
add: Using metal sheets on reagent dispensers will now produce a plumbing-enabled supply tank for station use.
add: Added a new stock part, the water recycler, used in building showers and sinks from scratch.
add: Hydroponics trays are now plumbing enabled. Water and nutrients will be sorted into their appropriate slots when piped in.
tweak: Sinks and showers now have internal water reservoirs, and start filled with water.
tweak: Showers and sinks must be constructed from a frame, then finished with a water recycler.
remove: Hydroponics hoses have been removed in favor of actual plumbing between machines.
/:cl:

### CODE REVIEW

**Comment by trexter555** [comment]
based

**Comment by necromanceranne** [comment]
My god, the future is now

**Comment by IndieanaJones** [comment]
Time to poison the water hole.

**Comment by TemporalOroboros** [comment]
> Make disconnected sinks/showers slowly regenerate water so that a stationwide plumbing net isn't required so Anturk doesn't snap my kneecaps off

Alternatively you could go all-in and give atmos a sister sub-department.

**Comment by LemonInTheDark** [comment]
slippery slope to fluidcode, but this isn't so bad

**Comment by Tharcoonvagh** [comment]
Add a subtype of plumbing pipes that run under the flooring

**Comment by Tharcoonvagh** [comment]
Like atmos and disposal pipes

**Comment by Time-Green** [comment]
> Add a subtype of plumbing pipes that run under the flooring

Ducts already do that, it works the same as everything else.

**Comment by OmegaDarkPotato** [comment]
Although I like the idea of being able to *change* what showers (and probably sinks) can pour out, I don't like the idea of having to worry about getting water pumped into them to have their base functionality. I'm all for cooperation that benefits both parties, but forcing cooperation to get your water to run isn't exactly appealing. I know that water isn't exactly super-precious right now, but we have machines that literally make matter out of energy (see: chem dispensers or even the booze dispenser) so why can't we have our sinks able to generate water for free?


Also, I think you have a mistake in your description where you write that "the shower drains reagents much faster than the shower", where I assume you meant "the shower drains reagents much faster than the sink".

**Comment by MrDoomBringer** [comment]
i'm personally all for making showers, sinks, etc. rely completely on water from plumbing. Gives the system more meaning and importance - I want to be scared when some villain threatens to poison the water line or whathave you.

**Comment by TemporalOroboros** [comment]
Oh dear god, I just realized. What happens if an antagonist virologist gets access to the water source? That could be _amazing_!

File: code/game/objects/structures/watercloset.dm [hunk 1]
```diff
@@ -278,6 +298,9 @@
 		return
 
 	busy = FALSE
+	reagents.remove_any(4)
+	reagents.expose(user, TOUCH)
+	reagents.trans_to(user, 1)
```
Comment (Jared-Fogle) [inline code/game/objects/structures/watercloset.dm:303]: Isn't this a no no? https://github.com/tgstation/tgstation/pull/52019

File: code/modules/reagents/reagent_dispenser.dm [hunk 2]
```diff
@@ -18,7 +29,18 @@
...
 /obj/structure/reagent_dispensers/attackby(obj/item/W, mob/user, params)
 	if(W.is_refillable())
-		return 0 //so we can refill them via their afterattack.
+		return FALSE //so we can refill them via their afterattack.
+	if(istype(W, /obj/item/stack/sheet/metal) && can_be_tanked)
+		var/obj/item/stack/sheet/metal/metal_stack = W
+		metal_stack.use(1)
+		var/obj/structure/reagent_dispensers/plumbed/storage/new_tank = new /obj/structure/reagent_dispensers/plumbed/storage(drop_location())
+		new_tank.reagents.maximum_volume = reagents.maximum_volume
+		reagents.trans_to(new_tank, (reagents.total_volume))
```
Comment (Jared-Fogle) [inline code/modules/reagents/reagent_dispenser.dm:38]:
```suggestion
		reagents.trans_to(new_tank, reagents.total_volume)
```

File: code/game/objects/items/stacks/sheets/sheet_types.dm [hunk 3]
```diff
@@ -90,7 +90,8 @@ GLOBAL_LIST_INIT(metal_recipes, list ( \
 	new/datum/stack_recipe("floodlight frame", /obj/structure/floodlight_frame, 5, one_per_turf = TRUE, on_floor = TRUE), \
 	new/datum/stack_recipe("voting box", /obj/structure/votebox, 15, time = 50), \
 	new/datum/stack_recipe("pestle", /obj/item/pestle, 1, time = 50), \
-	new/datum/stack_recipe("hygienebot assembly", /obj/item/bot_assembly/hygienebot, 2, time = 50)
+	new/datum/stack_recipe("hygienebot assembly", /obj/item/bot_assembly/hygienebot, 2, time = 50), \
```
Comment (Rohesie) [inline code/game/objects/items/stacks/sheets/sheet_types.dm:93]:
```suggestion
	new/datum/stack_recipe("hygienebot assembly", /obj/item/bot_assembly/hygienebot, 2, time = 5 SECONDS), \
```

File: code/game/objects/items/stacks/sheets/sheet_types.dm [hunk 4]
```diff
@@ -90,7 +90,8 @@ GLOBAL_LIST_INIT(metal_recipes, list ( \
...
+	new/datum/stack_recipe("shower frame", /obj/structure/showerframe, 2, time= 20)
```
Comment (Rohesie) [inline code/game/objects/items/stacks/sheets/sheet_types.dm:94]:
```suggestion
	new/datum/stack_recipe("shower frame", /obj/structure/showerframe, 2, time= 2 SECONDS)
```

File: code/datums/components/plumbing/_plumbing.dm [hunk 5]
```diff
@@ -231,13 +231,26 @@
...
+/datum/component/plumbing/simple_demand/hydroponics
+	var/datum/reagent/filtered_chem = /datum/reagent/water
+
+/datum/component/plumbing/simple_demand/hydroponics/process_request(amount, reagent, dir)
+	if(istype(parent, /obj/machinery/hydroponics))
+		var/obj/machinery/hydroponics/parent_tray = parent
+		if(reagent == filtered_chem && parent_tray.waterlevel <= parent_tray.maxwater)
+			parent_tray.adjustWater(round(amount))
+			reagents.remove_reagent(reagent, amount)
+	. = ..()
```
Comment (Rohesie) [inline code/datums/components/plumbing/_plumbing.dm:245]:
```suggestion
	return ..()
```

**Comment by Qustinnus** [comment]
cool pr. one thing to maybe consider though is that washing code should probably come from water application rather than showering itself. doesnt make sense that frying oil will wash blood off of you.

**Comment by Helianthus00** [comment]
It's an interesting idea, but I can't help but think this will be a nightmare on the mapping side. How do these stack layerwise with other pipes? Are we really going to have water pipes, disposals, scrubbers, air supply, and cable all on the same tile?

**Comment by Time-Green** [comment]
> It's an interesting idea, but I can't help but think this will be a nightmare on the mapping side. How do these stack layerwise with other pipes? Are we really going to have water pipes, disposals, scrubbers, air supply, and cable all on the same tile?

Showers and sinks autofill aswell. They aren’t connected by ducts until someone connects them by ducts in-game.

**Comment by Helianthus00** [comment]
> Showers and sinks autofill aswell. They aren’t connected by ducts until someone connects them by ducts in-game.

I see. I should have read more carefully. However, I still think certain areas will need this plumbing at roundstart, namely the kitchen and permabrig. Probably engineering showers as well, since they always leave them on for too long. It's better than needing the entire station to be covered in pipes, but I'm still concerned about running pipes alongside other utilities.

**Comment by TemporalOroboros** [comment]
We need a way to add ceiling and wall based utilities....

**Comment by MrDoomBringer** [comment]
plumbing should absolutley be mapped roundstart. If you really don't like the clutter, just make pipes only show up when holding a plumbing tool or mesons or whatever. No traitor is going to spend the first 45 minutes of their round adding ducts to all the showers and sinks on the map so they can poison the water supply, especially when most people probably wouldn't notice anyways

**Comment by TemporalOroboros** [comment]
I think we need ways to access piping/wiring/plumbing/etc in walls and possibly on ceilings, we're running out of space to put utilities in the floor tiles.

**Review by Twaticus — APPROVED** [review]

File: code/datums/components/plumbing/_plumbing.dm [hunk 6]
```diff
@@ -234,11 +234,36 @@
...
+	var/datum/reagent/filtered_chem = /datum/reagent/water
+	///typecast reference to our parent, which should always be a hydroponics tray.
+	var/obj/machinery/hydroponics/constructable/parent_tray
+
+/datum/component/plumbing/simple_demand/hydroponics/Initialize(start, _turn_connects)
+	if(istype(parent, /obj/machinery/hydroponics/constructable))
+		parent_tray = parent
+	. = ..()
+
+/datum/component/plumbing/simple_demand/hydroponics/transfer_to(datum/component/plumbing/target, amount, reagent, datum/ductnet/net)
```
Comment (ArcaneMusic) [inline code/datums/components/plumbing/_plumbing.dm:250]: This right here is breaking, gone through a few revisions of this now but no avail.

File: code/modules/reagents/reagent_dispenser.dm [hunk 7]
```diff
@@ -260,3 +260,9 @@
 	if(anchored)
 		to_chat(user, "<span class='warning'>It is fastened to the floor!</span>")
 	return !anchored
+
+/obj/structure/reagent_dispensers/plumbed/fuel
+	name = "stationairy fuel tank"
```
Comment (Jared-Fogle) [inline code/modules/reagents/reagent_dispenser.dm:265]: stationary

**Comment by cacogen** [comment]
i've wanted this for like 8 years or something

File: code/game/objects/items/storage/boxes.dm [hunk 8]
```diff
@@ -1386,3 +1386,15 @@
...
+	name = "box of plumbing supplies"
+	desc = "Contains a small supply of pipes, water recyclers, and metal to connect to the rest of the station."
+
+/obj/item/storage/box/plumbing/PopulateContents()
+	var/static/items_inside = list(
+		/obj/item/stock_parts/water_recycler = 2,\
+		/obj/item/stack/ducts/fifty = 1,\
+		/obj/item/stack/sheet/metal/ten = 1
+		)
+	generate_items_inside(items_inside,src)
```
Comment (Rohesie) [inline code/game/objects/items/storage/boxes.dm:1400]: don't really think this needs to be a static list
```suggestion
	var/list/items_inside = list(
		/obj/item/stock_parts/water_recycler = 2,
		/obj/item/stack/ducts/fifty = 1,
		/obj/item/stack/sheet/metal/ten = 1,
		)
	generate_items_inside(items_inside, src)
```

**Comment by Rohesie** [comment]
Want it TMed before full merge?

**Comment by ArcaneMusic** [comment]
I think a TM would probably be healthy, just to get more hands on it and get some preliminary ideas about the functionality.

**Comment by Rohesie** [comment]
I can only slap the label, you'll need to poke the keyholders for the actual TM.
Good luck!

**Comment by msgerbs** [comment]
Can we make it so the plumbing connections don't have to be only on the front?

**Comment by ArcaneMusic** [comment]
Travis is throwing a fit, but this works and had no issues in testmerge,

**Comment by LemonInTheDark** [comment]
I kicked his nuts in, should be good now

**Comment by LemonInTheDark** [comment]
This seems real neat, I hope you build it out in the future :)

**Comment by LemonInTheDark** [comment]
image

**Pull request merged**